File: 039/cpp/main.cpp
class Solution {
public:
    vector<vector<int>> solve(vector<int>& candidates, int target, int i) {
        if (i >= candidates.size()) {
            vector<vector<int>> t;
            if (target == 0) {
                t.push_back({});
            }
            return t;
        }
        int c = candidates[i];
        int r = target / c;
        vector<vector<int>> res;
        for (int k = 0; k < r + 1; k++) {
            vector<vector<int>> res0 = solve(candidates, target - k * c, i + 1);
            
            for (vector<int> &re0: res0) {
                for (int n = 0; n < k; n++) {
                    re0.insert(re0.begin(), c);
                }
                res.push_back(re0);
            }
        }
        return res;
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        return solve(candidates, target, 0);
    }
};

File: 039/rust/main.rs
impl Solution {
    fn solve(candidates: &Vec<i32>, target: i32, i: usize) -> Vec<Vec<i32>> {
        if i >= candidates.len() {
            return if target != 0 { Vec::new() } else { vec![vec![]] }
        }

        let c = candidates[i];
        let r = target / c;
        let mut res = Vec::new();
        for k in 0..r + 1 {
            let res0 = Solution::solve(candidates, target - k * c, i + 1);
            for re0 in res0 {
                let mut re = re0.clone();
                for n in 0..k {
                    re.insert(0, c);
                }
                res.push(re);
            }
        }
        return res;
    }

    pub fn combination_sum(candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        return Solution::solve(&candidates, target, 0);
    }
}
